back-end/routes/recommendations.js

```diff
@@ -18,34 +18,99 @@ const getAccessToken = async (userId) => {
     return token ? token.token : null;
 };
 
-const fetchStitchAverages = async (songs, access_token) => {
+const fetchAudioFeatures = async (songs, access_token) => {
     const songIds = songs.map(song => (song.uri).split(':').pop());
     const stringOfIds = songIds.join(',');
+
     try {
         const response = await fetch(`https://api.spotify.com/v1/audio-features?ids=${encodeURIComponent(stringOfIds)}`, {
             headers: {
                 'Authorization': `Bearer ${access_token}`
             }
         });
+        if (!response.ok) {
+            throw new Error(`HTTP error! status: ${response.status}`);
+        }
         const audioData = await response.json();
+        return audioData.audio_features;
     } catch (error) {
         console.error(error);
+        return [];
     }
 };
 
+function calculateGlobalFeatureAverages(features) {
+    let totalValence = 0;
+    let totalEnergy = 0;
+    let totalInstrumentalness = 0;
+    let totalCount = 0;
+
+    features.forEach(feature => {
+        totalValence += feature.valence;
+        totalEnergy += feature.energy;
+        totalInstrumentalness += feature.instrumentalness;
+        totalCount++;
+    });
+
+    const valence = totalCount > 0 ? totalValence / totalCount : 0;
+    const energy = totalCount > 0 ? totalEnergy / totalCount : 0;
+    const instrumentalness = totalCount > 0 ? totalInstrumentalness / totalCount : 0;
+
+    return {
+        valence,
+        energy,
+        instrumentalness
+    };
+}
+
+async function fetchAndProcessAudioFeatures(songs, access_token) {
+    const features = await fetchAudioFeatures(songs, access_token);
+    const averages = calculateGlobalFeatureAverages(features);
+    const lastSongFeatures = features.length > 0 ? features[features.length - 1] : null;
+    return {
+        averages,
+        lastSongFeatures
+    };
+}
+
+function gradingAlgorithm(lastSongFeatures, recommendationAudioFeatures, averages) {
```

Comment:
Function names should be verbs or verb phrases that explain what the function does, something like `calculateAndSortScores ` would make more sense here.

https://workat.tech/machine-coding/tutorial/writing-meaningful-variable-names-clean-code-za4m83tiesy0#:~:text=Function%20names%20should%20be%20verbs,a%20comment%20to%20describe%20it.